vispy/visuals/image_complex.py

```diff
@@ -0,0 +1,96 @@
+from .image import ImageVisual, _apply_clim_float, _apply_gamma_float
+import numpy as np
+from .shaders import Function, FunctionChain
+
+# In a complex Image, the texture will be rg32f, where:
+# data.r contains the real component
+# data.g contains the imaginary component
+COMPLEX_TRANSFORMS = {
+    "real": "float cplx2float(vec4 data) { return data.r; }",
+    "imaginary": "float cplx2float(vec4 data) { return data.g; }",
+    "magnitude": "float cplx2float(vec4 data) { return length(data); }",
+    "phase": "float cplx2float(vec4 data) { return atan(data.g, data.r); }",
+}
+# aliases
+COMPLEX_TRANSFORMS["angle"] = COMPLEX_TRANSFORMS["phase"]
+COMPLEX_TRANSFORMS["abs"] = COMPLEX_TRANSFORMS["magnitude"]
+COMPLEX_TRANSFORMS["imag"] = COMPLEX_TRANSFORMS["imaginary"]
+
+
+class ComplexImageVisual(ImageVisual):
+    COMPLEX_MODES = set(COMPLEX_TRANSFORMS)
+
+    def __init__(self, data=None, complex_mode="magnitude", **kwargs):
+        if complex_mode not in self.COMPLEX_MODES:
+            raise ValueError(
+                "complex_mode must be one of %s" % ", ".join(self.COMPLEX_MODES)
+            )
+        self._data_is_complex = np.iscomplexobj(data)
+        self._complex_mode = complex_mode
+
+        if kwargs.get("clim", "auto") == "auto" and self._data_is_complex:
+            kwargs["clim"] = self._calc_complex_clim(data)
+
+        kwargs["texture_format"] = "auto"
+        super().__init__(data=data, **kwargs)
+
+    def set_data(self, image):
+        data = np.asarray(image)
+        self._data_is_complex = np.iscomplexobj(data)
+        if self._data_is_complex:
+            #  Turn the texture into an rg32f texture
+            # where r = 'real' and g = 'imag'
+            self._texture._format = "rg"
+            data = np.stack([data.real, data.imag], axis=-1)
+        else:
+            self._texture._format = None
+        return super().set_data(data)
+
+    @property
+    def complex_mode(self):
+        return self._data_is_complex and self._complex_mode
+
+    @complex_mode.setter
+    def complex_mode(self, value):
+        if value not in self.COMPLEX_MODES:
+            raise ValueError(
+                "complex_mode must be one of %s" % ", ".join(self.COMPLEX_MODES)
+            )
+        if self._complex_mode != value:
+            self._complex_mode = value
+            self._need_colortransform_update = True
+            self.update()
+
+    def _build_color_transform(self):
+        if self.complex_mode:
```

Comment:
It is impossible for this to be False right? So this could be removed?

Reply:
well, with that line... the `ComplexImageVisual` should technically be possible to present both complex and non-complex data.  (i.e. it's effectively the same as the more class that we had before the separation).  If someone were to call `set_data` with non-complex data, then `self._data_is_complex` would go back to False, and then the `complex_mode` property would return `False`... I _think_ it might mean that napari could simply use the complex version and support all uses cases (I haven't thoroughly tested that, and will write some tests to see... but that was the motivation)

Reply:
Added a test to confirm that complexImage can also be used for non complex data

Reply:
Thanks.